lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -1990,56 +2063,33 @@ bool EarlyFunctionAnalysis::runOnModule(Module &M) {
   Properties.importModel();
 
   // Instantiate a FunctionEntrypointAnalyzer object
-  FEA Analyzer(M, &GCBI, ABIRegisters, Properties, Binary);
+  FEA Analyzer(M, &GCBI, ABIRegisters, &EntrypointsQueue, Properties, Binary);
 
-  // Interprocedural analysis over the collected functions in post-order
-  // traversal (leafs first).
-  while (!EntrypointsQueue.empty()) {
-    BasicBlockNode *EntryNode = EntrypointsQueue.pop();
-    revng_log(EarlyFunctionAnalysisLog,
-              "Analyzing Entry: " << EntryNode->BB->getName());
-
-    // Intraprocedural analysis
-    FunctionSummary AnalysisResult = Analyzer.analyze(EntryNode->BB);
-    bool Changed = Properties.registerFunction(getBasicBlockPC(EntryNode->BB),
-                                               std::move(AnalysisResult));
-
-    // If we got improved results for a function, we need to recompute its
-    // callers, and if a caller turns out to be fake, the callers of the fake
-    // function too.
-    if (Changed) {
-      UniquedQueue<BasicBlockNode *> FakeFunctionWorklist;
-      FakeFunctionWorklist.insert(EntryNode);
+  // Either recover the control-flow graph of the program and serialize it to
+  // LLVM IR, or schedule a full ABI analysis for each function detected
+  // (control-flow graph recovery included).
+  if (!ShouldAnalyzeABI) {
+    Analyzer.recoverCFG();
+  } else {
+    // Interprocedural analysis over the collected functions in post-order
+    // traversal (leafs first).
+    Analyzer.runInterproceduralAnalysis();
 
-      while (!FakeFunctionWorklist.empty()) {
-        BasicBlockNode *Node = FakeFunctionWorklist.pop();
-        for (auto *Caller : Node->predecessors()) {
-          if (Caller == RootNode)
-            break;
+    // Propagate results between call-sites and functions
+    Analyzer.interproceduralPropagation();
 
-          if (!Properties.isFakeFunction(getBasicBlockPC(Caller->BB)))
-            EntrypointsQueue.insert(Caller);
-          else
-            FakeFunctionWorklist.insert(Caller);
-        }
-      }
-    }
+    // Commit the results onto the model
+    Analyzer.finalizeModel();
   }
 
+  // Serialize function metadata, including the CFG, to the entry basic blocks
+  // in `root`.
+  serializeFunctionMetadata(M.getContext(), GCBI, Properties, Binary);
+
```

Comment:
I think we should move this into the `if (!ShouldAnalyzeABI)` branch and run `opt -detect-abi -collect-cfg`.
We're duplicating some work, but we have a nice property:
* `-detect-abi` writes the model but not the IR.
* `-collect-cfg` writes the IR but not the model.
Would this work?

Reply:
Is it really a demanding property that we want to have? I ask this because this would imply to always run `--detect-abi` alongside `--collect-cfg`, if we want to write both on the IR and on the model, and I don't think we should want to have this distinction at this granularity. AFAIC, EFA has been designed to do both, but doing so, we would split the two things at the extent in which part of the work would be duplicated (as you noted). Furthermore, as far as we conceived this, CFG collection-only should be a subset of what `--detect-abi` does, thus --detect-abi` should be allowed to write the IR as well.

Reply:
> Is it really a demanding property that we want to have?

Yes. I really want a clear distinction between those writing the model and those writing the IR.

> I ask this because this would imply to always run --detect-abi alongside --collect-cfg

No, you're still thinking in terms of one linear pipeline. But `--detect-abi` could be run on its own just to produce effects on the model and then not run anything after that.

> part of the work would be duplicated

This is just a performance issue that we can tackle in case it becomes an actual bottleneck.

> CFG collection-only should be a subset of what --detect-abi does, thus --detect-abi` should be allowed to write the IR as well

It's a subset only in terms of the computations it makes, not in terms of where it writes its results.